designs/function-node-lifecycle/README.md

```diff
@@ -41,17 +41,100 @@ Following figure shows new lifecycle model of function node.  While creating a f
+![config](./config.png)
+
+A user can use required NPM module through a variable defined in this setting panel.
+
+For example, with above setting, `fs` variable in function node code can access `fs-ext` module without `require` call.
+
+If a user specifies different versions of NPM module, version conflicts may occur.  The Node-RED runtime provides a feature for installing different versions.
+
+If Projects feature is enabled, the Node-RED runtime updates project's `package.json` on deploy if specified in `settings.js`.
+
+#### Additional settings for external npm module in `settings.js`
+
+Add a top-level property named `externalModule` to `settings.js`.  It points to a object with following properties.
+
+| name | type   | required | description  |
+| :--- | :----- | :------- | :----------- |
+| mode | string | yes      | one of `"auto"`, `"auto-update"`, `"manual"`, or `"none"` (default: `"manual"`)<br/>- `auto`: automatically install requested NPM module if not installed.<br/>- `auto-update`: same as `auto` but with automatic update if newer version is available.<br/>- `manual`: users can manually install requested NPM module from user settings panel.<br/>- `none`: installation of external NPM modules is not allowd |
+| allowList | array of string | no | List of regular expressions that matches module's NPM registry specification. External libraries that matches this list **can** be used in `Function` nodes. |
+| denyList | array of string | no | List of regular expressions that matches module's NPM registry specification. External libraries that matches this list **can not** be used in `Function` nodes. |
+
+`allowList` precedes over `denyList`.
+
+```
+    // Example of settings.js
+    externalModules: {
+        mode: "manual",
+        allowList: [
+            "^fs-ext$",       // allow fs-ext
+            "^qrcode@1.4.4$"  // allow qrcode version 1.4.4
+        ],
+        denyList: [
+            ".*"              // disallow others 
+        ]
+    },
+```
+If the `externalModule` property not exists, `none` mode and denyList of `".*"`, meaning not allow installation and use of external module, is expected as a default.
+
+#### New API for installing NPM module
+
+In order to allow NPM modules to be installed automatically before node execution, `RED.nodes.registerType` will be extended to accept new `dynamicModuleList` property in an optional third parameter.  It may have following value:
+
+1.  a `string` value that represents a property name of a node instance.  The property of the node instance must have a list of modules to install.
+
+   ```
+   // Node definition
+   RED.nodes.registerType("function", FunctionNode, {
+       dynamicModuleList: "modules"
+   });
+   ```
+
+   ```
+   // Node instance definition
+   {
+       ...
+       "modules": [ "fs-ext@1.2.3", "qrcode@1.4.4" ]
+       ...
+   }
+   ```
+
+2.  a function that accepts node instance definition and returns a list of modules to install.
```

Comment:
How about we say the value of the property `dynamicModuleList` points at can be either:

1. an array of Strings: `["fs-ext@1.2.3", "qrcode@1.4.4" ]`
2. or an array of Objects that must have a `name` property (but can have any other properties): `[ { "name": "fs-ext@1.2.3", "var": "fs", ... }, { "name": "qrcode", ...... }]`

Does that give enough flexibility without requiring the node author to have to also provide a mapping function?

We could always add support for a mapping function in the future if it was absolutely necessary, but I'm not sure its needed now.

Reply:
Changed to simple array of objects according to your suggestion. 
Added a couple of fixes related to local module installations.